File: scala/digital-wallet/src/main/scala/core/services/WalletsService.scala
package core.services

import core.domain.entities.{Transaction, Wallet}
import core.domain.enums.*
import core.domain.enums.SubwalletType.SubwalletType
import core.domain.model.*
import core.errors.{InvestmentFailedError, LiquidationFailedError}
import ports.{InvestmentPolicyDatabase, WalletFilter, WalletsDatabase}

class WalletsService(walletsRepo: WalletsDatabase, investmentPolicyRepo: InvestmentPolicyDatabase, ledgerService: LedgerService, transactionsService: TransactionsService, investmentService: InvestmentService) {
  def getAvailableBalance(wallet: Wallet): BigDecimal = {
    val ledgerQuery = wallet.walletType match {
      case WalletType.RealMoney =>
        List(
          LedgerQuery(
            subwalletType = SubwalletType.RealMoney,
            balanceType = BalanceType.Available
          )
        )
      case WalletType.Investment =>
        List(
          LedgerQuery(
            subwalletType = SubwalletType.Bonds,
            balanceType = BalanceType.Available,
          ),
          LedgerQuery(
            subwalletType = SubwalletType.Stock,
            balanceType = BalanceType.Available,
          ),
          LedgerQuery(
            subwalletType = SubwalletType.RealEstate,
            balanceType = BalanceType.Available,
          ),
          LedgerQuery(
            subwalletType = SubwalletType.Cryptocurrency,
            balanceType = BalanceType.Available,
          )
        )
      case WalletType.EmergencyFunds =>
        List(
          LedgerQuery(
            subwalletType = SubwalletType.EmergencyFunds,
            balanceType = BalanceType.Available
          )
        )
    }

    ledgerService.getBalance(wallet.id, ledgerQuery)
  }
  
  def getSubwalletPendingBalance(walletId: String, subwalletType: SubwalletType): BigDecimal = {
    ledgerService.getBalance(
      walletId,
      List(
        LedgerQuery(
          subwalletType = subwalletType,
          balanceType = BalanceType.Holding
        )
      )
    )
  }

  def invest(request: InvestmentRequest): Either[InvestmentFailedError, Transaction] = {
    val wallets = 
      walletsRepo.find(
        WalletFilter(
          customerId = Some(request.customerId), 
          walletType = Some(WalletType.RealMoney)
        )
      )

    wallets match {
      case List(wallet) =>
        for {
          transaction <- transactionsService.create(
            CreateTransactionRequest(
              amount = request.amount,
              idempotencyKey = request.idempotencyKey,
              originatorWalletId = wallet.id,
              originatorSubwalletType = SubwalletType.RealMoney,
              transactionType = TransactionType.Hold
            )
          ).left.map { e =>
            InvestmentFailedError(e.message)
          }
          
          processTransactionTuple <- 
            transactionsService.process(transaction).left.map { e =>
              transactionsService.updateStatus(transaction.id, TransactionStatus.Failed)
              InvestmentFailedError(e.message)
            }
            
          executedTransaction <- 
            transactionsService.execute(processTransactionTuple).left.map(e => 
            InvestmentFailedError(e.message)
          )
        } yield executedTransaction

      case _ =>
        Left(InvestmentFailedError(s"None or multiple wallets found for customer ${request.customerId}"))
    }
  }

  def liquidate(request: LiquidationRequest): Either[LiquidationFailedError, Unit] = {
    val wallets = walletsRepo.find(WalletFilter(customerId = Some(request.customerId), walletType = Some(WalletType.Investment)))

    wallets match {
      case List(wallet) =>
          for {
            policyId <- wallet.policyId.toRight(LiquidationFailedError(s"Wallet ${wallet.id} has no policyId"))
            
            investmentPolicy <-
              investmentPolicyRepo
                .findById(policyId)
                .toRight(LiquidationFailedError(s"Investment policy ${wallet.policyId} not found"))
            
            _ <- investmentService.executeMovementWithInvestmentPolicy(MovementRequest(
                  amount = request.amount,
                  idempotencyKey = request.idempotencyKey,
                  walletId = wallet.id,
                  investmentPolicy = investmentPolicy,
                  transactionType = TransactionType.Hold
                )).left.map(e => LiquidationFailedError(e.message))
          } yield ()
      case _ => Left(LiquidationFailedError(s"None or multiple wallets found for customer ${request.customerId}"))
    }
  }
}

File: kotlin/digital-wallet/src/main/kotlin/digitalwallet/core/services/WalletsService.kt
package digitalwallet.core.services

import digitalwallet.core.domain.enums.SubwalletType
import digitalwallet.core.domain.enums.TransactionStatus
import digitalwallet.core.domain.enums.TransactionType
import digitalwallet.core.domain.enums.WalletType
import digitalwallet.core.domain.models.InvestmentRequest
import digitalwallet.core.domain.models.LiquidationRequest
import digitalwallet.core.domain.models.ProcessTransactionRequest
import digitalwallet.core.exceptions.PartnerException
import digitalwallet.core.exceptions.TransactionFailed
import digitalwallet.core.exceptions.ValidationException
import digitalwallet.ports.InvestmentPolicyDatabase
import digitalwallet.ports.WalletFilter
import digitalwallet.ports.WalletsDatabase

class InvestmentFailed(
    message: String,
) : Exception()

class LiquidationFailed(
    message: String,
) : Exception()

class WalletsService(
    private val walletsRepo: WalletsDatabase,
    private val investmentPolicyRepo: InvestmentPolicyDatabase,
    private val transactionsService: TransactionsService,
    private val investmentService: InvestmentService,
) {
    suspend fun invest(request: InvestmentRequest) {
        val wallet =
            walletsRepo
                .find(
                    WalletFilter(
                        customerId = request.customerId,
                        type = WalletType.REAL_MONEY,
                    ),
                ).firstOrNull() ?: throw NoSuchElementException("Wallet not found for customer ${request.customerId}")

        val processTransactionRequest =
            ProcessTransactionRequest(
                amount = request.amount,
                idempotencyKey = request.idempotencyKey,
                originatorWalletId = wallet.id,
                originatorSubwalletType = SubwalletType.REAL_MONEY,
                type = TransactionType.HOLD,
            )

        try {
            transactionsService.processTransaction(processTransactionRequest)
        } catch (e: ValidationException) {
            println(e)
            transactionsService.handleException(e, TransactionStatus.FAILED, request.idempotencyKey)
            throw InvestmentFailed(e.message.toString())
        } catch (e: PartnerException) {
            transactionsService.handleException(e, TransactionStatus.TRANSIENT_ERROR, request.idempotencyKey)
        }
    }

    suspend fun liquidate(request: LiquidationRequest) {
        val wallet =
            walletsRepo
                .find(
                    WalletFilter(
                        customerId = request.customerId,
                        type = WalletType.INVESTMENT,
                    ),
                ).firstOrNull() ?: throw NoSuchElementException("Wallet not found for customer ${request.customerId}")
        val investmentPolicy =
            investmentPolicyRepo.findById(wallet.policyId) ?: throw NoSuchElementException("Policy ${wallet.policyId} not found")

        try {
            investmentService.executeMovementWithInvestmentPolicy(
                InvestmentMovementRequest(
                    amount = request.amount,
                    idempotencyKey = request.idempotencyKey,
                    walletId = wallet.id,
                    investmentPolicy = investmentPolicy,
                    transactionType = TransactionType.HOLD,
                ),
            )
        } catch (e: TransactionFailed) {
            throw LiquidationFailed(e.message.toString())
        }
    }
}
